### ISSUE TITLE
Distributed Testing Should Not Use a Distributed Sampler

### ISSUE THREAD

**Issue by Queuecumber** [PyTorchLightning/pytorch-lightning#7929]
## 🐛 Bug

This is a fairly important bug report that I've been meaning to make for a while. 

In general, it is incorrect to try to do **testing** with a distributed sampler. This is because the distributed sampler is either going to mix in already processed samples or drop samples in order to make the number of batches divide evenly on the number of GPUs.

This is fine when you're doing training because you just want a bunch of data and you arent too picky about it, and its **usually** fine when you're doing validation and you just want to get a sense of the model performance.

But for **testing** those are numbers that are potentially going into your paper which you're going to report to the world, and if distributed sampler was repeating or dropping samples, your dataset averages are going to be incorrect. So basically PL is, by default, contributing to the reporting of incorrect results. 

To make matters worse, there's no easy way to opt out of this. You need to tel PL not to replace *any* samplers, then add the samplers for train and val yourself. 

The fix for this is to make distributed testing opt-in and false by default. I recommend changing the `replace_sampler_ddp` parameter to take a string or enum in addition to bool, it can be something like

|`replace_sampler_ddp` value | meaning|
| - | - |
| `False` | No distributed samplers |
| `True` | All distributed sampler |
| `train` | Only training dataloader gets a distributed sampler |
| `train_val` | Train and val DLs get distributed samplers |

with the default being `train_val`

### Additional context

As always I'm happy to PR this

More context in a slack discussion here (no idea if that link will work): 

https://pytorch-lightning.slack.com/archives/CRBLFHY79/p1622595393383300

Note that the larger the dataset the more of an effect this will have (and the few GPUs)

I think this is a very important bug to get fixed because most people won't notice it, their results will just be off and they'll never know. This can make results artificially better or worse.

**Comment by Queuecumber** [comment]
Actually just following up on something I read on torchmetrics, would it be possible to incorporate [join](https://pytorch.org/docs/stable/generated/torch.nn.parallel.DistributedDataParallel.html#torch.nn.parallel.DistributedDataParallel.join) into PLs loop somehow?

**Comment by Queuecumber** [comment]
Yeah I can PR the proposed API changes if you guys are interested in that approach, that part is simple. The join idea is interesting but it would require some investigation to get it right

**Comment by tchaton** [comment]
Dear @Queuecumber,

Here are my recommendation for this issue:
1. Let's open a PR to improve documentation and add a section `paper submissions with Lightning` and add tips around the problem you just raised

2. I believe  `replace_sampler_ddp` is more a problem for `test` than `validation` as it is usually the dataset used for paper submission. 
Small note: In the case where no distributed sampler is used in a distributed setting, TorchMetrics Metric shouldn't be synced as entire dataset is processed by all processes.  

3. Ultimately, I believe supporting `join` for `validation` and `test` would definitely be the safest solution.

Best,
T.C

**Comment by Queuecumber** [comment]
> Dear @Queuecumber,
> 
> Here are my recommendation for this issue:
> 
>     1. Let's open a PR to improve documentation and add a section `paper submissions with Lightning` and add tips around the problem you just raised
Yeah we should have this in the docs, I also recommend the proposed API change as it makes it possible to turn of test-time auto distributed sampler, there's currently no easy way to do that. 
> 
>     2. I believe  `replace_sampler_ddp` is more a problem for `test` than `validation` as it is usually the dataset used for paper submission.
>        Small note: In the case where no distributed sampler is used in a distributed setting, TorchMetrics Metric shouldn't be synced as entire dataset is processed by all processes.
Correct
> 
>     3. Ultimately, I believe supporting `join` for `validation` and `test` would definitely be the safest solution.
> 
I'd love to see something like this but I have no idea how to implement it
> 
> Best,
> T.C

I should be able to make the PR this week

**Comment by ananthsub** [comment]
For reference, here's the prior issue for DDP join: https://github.com/PyTorchLightning/pytorch-lightning/issues/3325

however for testing, why do we need DDP at all since there are no grads that need to be synced

**Comment by Queuecumber** [comment]
Ok so it sounds like join may not work out any time soon

> however for testing, why do we need DDP at all since there are no grads that need to be synced

Correct, so join isn't even necessary here, but there's still a distributed sampler in play and metric syncing would been to be handled right? e.g, if process A had 4 batches and process B had 5 then process B would hang trying to sync it's metrics
